core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1935,4 +2040,122 @@ object Log {
+          triesSoFar += 1
+          if (triesSoFar == maxTries)
+            throw e
+        }
+      }
+    }
+    throw new IllegalStateException()
+  }
+
+  /**
+   * Split the given log segment into multiple such that there is no offset overflow in the resulting segments. The
+   * resulting segments will contain the exact same messages that are present in the input segment. On successful
+   * completion of this method, the input segment will be deleted and will be replaced by the resulting new segments.
+   * See replaceSegments for recovery logic, in case the broker dies in the middle of this operation.
+   * Note that this method assumes we have already determined that the segment passed in contains records that cause
+   * offset overflow.
+   * @param log The log containing segment to split
+   * @param segment Segment to split
+   * @return List of new segments that replace the input segment
+   */
+  private[log] def splitSegmentOnOffsetOverflow(log: Log, segment: LogSegment): List[LogSegment] = {
+    val newSegments = ListBuffer[LogSegment]()
+    var position = 0
+    val sourceRecords = segment.log
+    var readBuffer = ByteBuffer.allocate(1024 * 1024)
+
+    class CopyResult(val bytesRead: Option[Int], val overflowOffset: Option[Long])
+
+    // Helper method to copy `records` into `segment`. Makes sure records being appended do not result in offset overflow.
+    def copyRecordsToSegment(records: FileRecords, segment: LogSegment, readBuffer: ByteBuffer): CopyResult = {
+      val baseOffset = segment.baseOffset
+      var bytesRead = 0
+      var maxTimestamp = Long.MinValue
+      var offsetOfMaxTimestamp = Long.MinValue
+      var maxOffset = Long.MinValue
+
+      // find all batches that are valid to be appended to the current log segment
+      val (validBatches, overflowBatches) = records.batches.asScala.span(batch => (batch.lastOffset - Int.MaxValue <= baseOffset))
+      val overflowOffset = overflowBatches.headOption.map(firstBatch => {
+        log.info(s"Found overflow at offset ${firstBatch.baseOffset} in segment $segment of log $log")
+        firstBatch.baseOffset
+      })
+
+      // return early if no valid batches were found
+      if (validBatches.size == 0)
+        return new CopyResult(None, overflowOffset)
+
+      // read all the valid batches
+      val lastValidBatch = validBatches.last
+      val validRecords = records.slice(0, lastValidBatch.position + lastValidBatch.sizeInBytes)
+      require(readBuffer.capacity >= validRecords.sizeInBytes)
+      readBuffer.clear()
+      readBuffer.limit(validRecords.sizeInBytes)
+      validRecords.readInto(readBuffer, 0)
+
+      for (batch <- validBatches) {
+        if (batch.maxTimestamp > maxTimestamp) {
+          maxTimestamp = batch.maxTimestamp
+          offsetOfMaxTimestamp = batch.lastOffset
+        }
+
+        maxOffset = batch.lastOffset
+        bytesRead += batch.sizeInBytes
```

Comment:
`bytesRead` is what we need to compute the slice size, right? Maybe we can move the block starting from line 2100 below this method and then we don't need `lastValidBatch`.